broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/NewPartitionTransitionImpl.java

```diff
@@ -71,67 +73,48 @@ public void updateTransitionContext(final PartitionTransitionContext transitionC
     steps.forEach(step -> step.onNewRaftRole(context, role));
 
     final ActorFuture<Void> nextTransitionFuture = concurrencyControl.createFuture();
+    final var nextTransition =
+        new PartitionTransitionProcess(steps, concurrencyControl, context, term, role);
+    nextTransitionFuture.onComplete((v, e) -> lastTransition = nextTransition);
+
+    final var ongoingTransitionFuture =
+        currentTransitionFuture == null
+            ? concurrencyControl.createCompletedFuture()
+            : currentTransitionFuture;
+
+    // For safety reasons we have to immediately replace the current transition future with the next
+    // transition future, such that we make sure that we enqueue all transitions after another.
+    //
+    // This means we will always add the next transition to the tail, this can become a chain of
+    // futures if we have many transitions after another. Ideally we would execute after the current
+    // transition only the last enqueue one. This should be implemented as soon we have time for it.
+    currentTransitionFuture = nextTransitionFuture;
 
-    concurrencyControl.run(
-        () -> {
-          if (currentTransition != null) {
-            LOG.info(
-                format(
-                    "Transition to %s on term %d requested while another transition is still running",
-                    role, term));
-            currentTransition.cancel(); // this will drop any subsequent transition steps
-
-            // schedule new transition as soon as the current step of the current transition
-            // has completed
-            concurrencyControl.runOnCompletion(
-                currentTransitionFuture,
-                (ok, error) -> cleanupLastTransition(nextTransitionFuture, term, role));
+    ongoingTransitionFuture.onComplete(
+        (nothing, error) -> {
+          context.setCurrentTerm(term);
+          context.setCurrentRole(role);
 
+          if (lastTransition == null) {
+            nextTransition.start(nextTransitionFuture);
           } else {
-            cleanupLastTransition(nextTransitionFuture, term, role);
+            final var cleanupFuture = lastTransition.cleanup(term, role);
+            cleanupFuture.onComplete(
+                (ok, e) -> {
+                  if (error != null) {
+                    LOG.error(
+                        String.format("Error during transition clean up: %s", error.getMessage()),
+                        error);
+                    LOG.info(
```

Comment:
:question: Another dumb question :upside_down_face:, since I know you just copied this, but maybe someone can tell me - why do we have an ERROR log statement and then an INFO one?

Reply:
The error reports the error. The info informs about what is going to happen next.